### DISCUSSION TITLE
Nested JSON

### DISCUSSION THREAD

**Discussion by FHJCM** [TomWright/dasel#438] [Q&A]
I am having trouble extracting fields from a Json file with three levels of nesting and converting them to CSV.

I would like the output to look like this:
Customer, Type of Work, ASO, Type of Work-Subtype, Amount
Acme Ltd, Managed Services, ASO-2197, Configuration, 20,100.00
The Power Company, Consultancy, ASO-2178, New, 6,300.00

The first three records of the JSON file are as follows.
`{"expand":"schidema,names",
"startAt":0,
"maxResults":100,
"total":92,
"issues":[
{
"expand":"operations,versionedRepresentations,editmeta,changelog,customfield_10134.properties,renderedFields",
"id":"17084",
"self":"https://automation-consultants.atlassian.net/rest/api/2/issue/17084",
"key":"ASO-2178",
"fields":{
"customfield_10140":
[{"self":"https://automation-consultants.atlassian.net/rest/api/2/customFieldOption/10532",
"value":"New",
"id":"10532"}],
"customfield_10139":{
"self":"https://automation-consultants.atlassian.net/rest/api/2/customFieldOption/10528",
"value":"Consultancy",
"id":"10528"},
"customfield_10111":6300.0,
"customfield_10134":{
"id":417,
"value":"The Power Company"}}},
{
"expand":"operations,versionedRepresentations,editmeta,changelog,customfield_10134.properties,renderedFields",
"id":"17149",
"self":"https://automation-consultants.atlassian.net/rest/api/2/issue/17149",
"key":"ASO-2197",
"fields":{
"customfield_10140":
[{"self":"https://automation-consultants.atlassian.net/rest/api/2/customFieldOption/10536",
"value":"Configuration",
"id":"10536"},
{"self":"https://automation-consultants.atlassian.net/rest/api/2/customFieldOption/10542",
"value":"ITSM",
"id":"10542"},
{"self":"https://automation-consultants.atlassian.net/rest/api/2/customFieldOption/10533",
"value":"Renewal",
"id":"10533"},
{"self":"https://automation-consultants.atlassian.net/rest/api/2/customFieldOption/10534",
"value":"Upgrade",
"id":"10534"}],
"customfield_10139":{
"self":"https://automation-consultants.atlassian.net/rest/api/2/customFieldOption/10529",
"value":"Managed Services",
"id":"10529"},
"customfield_10111":20100.0,
"customfield_10134":{
"id":76,"value":"Acme Ltd"}}},
{
"expand":"operations,versionedRepresentations,editmeta,changelog,customfield_10134.properties,renderedFields",
"id":"17628",
"self":"https://automation-consultants.atlassian.net/rest/api/2/issue/17628",
"key":"ASO-2319",
"fields":{
"customfield_10140":
[{"self":"https://automation-consultants.atlassian.net/rest/api/2/customFieldOption/10532",
"value":"New",
"id":"10532"}],
"customfield_10139":{
"self":"https://automation-consultants.atlassian.net/rest/api/2/customFieldOption/10528",
"value":"Consultancy",
"id":"10528"},
"customfield_10111":4500.0,
"customfield_10134":{
"id":417,"value":"The Power Company"}}},`

I can get the ASO and Amount columns with the following dasel command.
`dasel -f JQL_out0 -r json  -w csv 'issues.all().mapOf(ASO,key,Amount,fields.customfield_10111).merge()'`

But I can't get any deeper into the JSON than this, and I can't extract the Customer name., ,for example. Any help to fix this would be much appreciated.

**Comment by FHJCM** [comment]
I had it working in dasel V1 (see dasel V1 expression below), but dasel V2 has defeated me.

`dasel -p json -f JQL_out0 -m ".issues.[*]" --format '{{select ".fields.customfield_10134.value"}},{{select ".fields.customfield_10139.value"}},{{select ".key"}},{{select ".fields.customfield_10140.[0].value"}},{{select ".fields.customfield_10111"}}'`

**Comment by TomWright** [comment]
Something like this?
`'issues.all().mapOf(ASO,key,Amount,fields.customfield_10111,CustomerName,fields.customfield_10134.value).merge()'`

```json
[
  {
    "ASO": "ASO-2178",
    "Amount": 6300,
    "CustomerName": "The Power Company"
  },
  {
    "ASO": "ASO-2197",
    "Amount": 20100,
    "CustomerName": "Acme Ltd"
  },
  {
    "ASO": "ASO-2319",
    "Amount": 4500,
    "CustomerName": "The Power Company"
  }
]
```

**Reply by TomWright** [reply]
If you can give me the example output you want for the given input above I can put something together for you.

**Reply by fmiers** [reply]
Thank you very much for getting back to me. In the end, having had a couple of late nights and a deadline pressing, I had to give up and wrote a Python script instead. To answer your questions about format, any output format would have done, provided it was in CSV and contained third level data such as the Customer name.

It would be really helpful to have lots more documentation and worked examples than exist at the moment. I got to the point where I doubted it was even possible to extract third-level data using V2 of dasel. I did not even want to upgrade to V2 of dasel but had to because I was issued with a new laptop and had to reinstall dasel, which of course gave me V2 instead of the V1 I had been using. I tried installing V1 again, but it would not allow that. Thus, my code was broken and what should have been an hour at most to upgrade three dasel V1 expressions turned into multiple late nights of failure to port them to V2.

The main problems I encountered are:
1. The documentation is very thin, with few worked examples. The MapOf() function gets a paragrah and a few entry-level examples.
2. My doubt about V2 was caused by the fact that the mapOf() function fails if there are more than two layers of nesting. mapOf("heading",field.level1) works, but mapOf("heading", fleld.level1.level2) fails every time. Maybe there is a good reason for the MapOf() function only to allow 2 levels, but if so, why is that not explained in the documentation?
3. When dasel fails, the error messages are very generic and are of little help in diagnosing the problem.

I uploaded all the documentation into ChatGPT-4o, to see if it could make sense of it, but it could not, because there was not enough for the LLM to go on. Note that ChatGPT-4o is pretty good at generating Python code or Excel VBA. 

A working dasel would be very useful and there is definitely a need for it, but the current version effectively shortened my life by about six hours without delivering any value.
